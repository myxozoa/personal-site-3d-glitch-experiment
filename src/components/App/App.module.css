.app {
  display: flex;
  flex-direction: column;
  height: 100%;
  pointer-events: none;
}

.header {
  font-weight: 300;
  font-size: 2rem;
  letter-spacing: 1.5px;
  color: hsl(0, 0%, 98%);
}

.name {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 100px;
  font-size: 3rem;
  font-weight: 500;
  padding: 5px;
  margin-bottom: 20px;
}

.lastName {
  display: flex;
}

.periodContainer {
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  padding: 1px;
}

.period {
  height: 10px;
  width: 10px;
  background: white;
  border-radius: 50%;
}