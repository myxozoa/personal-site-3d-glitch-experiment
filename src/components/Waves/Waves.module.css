.container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  position: fixed;
  background: hsl(0, 0%, 2%);
  z-index: -100;
  top: 0;
  left: 0;
  pointer-events: auto;
}

.root {
  width: 100%;
  height: 100%;
}