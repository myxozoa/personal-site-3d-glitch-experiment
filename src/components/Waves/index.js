import React from 'react';
import * as THREE from 'three';

import styles from './Waves.module.css';

import vertShader from './particle.vert';
import fragShader from './particle.frag';
import particle from '../../assets/particle.png';
import Stats from '../../stats.min';

import { getMousePos } from '../../helpers/getMousePos';

class Wave extends React.Component {
  constructor(props) {
    super(props);
    const { width, height, separation, amountX, amountY, debug } = this.props;

    this.container = React.createRef();
    this.width = width;
    this.height = height;
    this.separation = separation;
    this.amountX = amountX;
    this.amountY = amountY;
    this.debug = debug;
    this.mouseX = 0;
    this.mouseY = 0;
    this.sizeMultiplier = 1.5;
    this.hover = false;
    this.jump = 0;

    this.clock = null;
    this.geometry = null;
    this.particleSystem = null;
    this.camera = null;
    this.scene = null;
    this.renderer = null;
    this.stats = null;
    this.material = null;
    this.originalCameraPos = {};
  }

  componentDidMount() {
    this.init();
    this.animate();
  }

  onMouseMove = event => {
    this.hover = true;
    const mousePos = getMousePos(event);
    this.mouseX = -(mousePos.x - this.width / 2) / 2;
    this.mouseY = -(mousePos.y - this.height / 2) * 2;
  };

  init = () => {
    this.container.current.addEventListener('mousemove', this.onMouseMove);

    this.width = this.container.current.clientWidth;
    this.height = this.container.current.clientHeight;

    this.sizeMultiplier = 1 + this.height / 450;

    this.camera = new THREE.PerspectiveCamera(75, this.width / this.height, 1, 10000);

    this.camera.aspect = this.width / this.height;

    this.originalCameraPos = {
      x: this.camera.position.x,
      y: 500 + this.height * 2,
      z: 2000 + (this.height + this.width),
    };

    this.camera.position.set(this.originalCameraPos.x, this.originalCameraPos.y, this.originalCameraPos.z);

    this.scene = new THREE.Scene();

    this.clock = new THREE.Clock();

    this.clock.start();

    this.camera.lookAt(this.scene.position);

    const uniforms = {
      texture: {
        value: new THREE.TextureLoader().load(particle),
      },
    };

    this.material = new THREE.ShaderMaterial({
      uniforms: uniforms,
      vertexShader: vertShader,
      fragmentShader: fragShader,
      blending: THREE.AdditiveBlending,
      depthTest: false,
      vertexColors: true,
    });

    this.geometry = new THREE.BufferGeometry();

    const color = new THREE.Color();
    const positions = [];
    const colors = [];
    const sizes = [];

    for (var ix = 0; ix < this.amountX; ix++) {
      for (var iy = 0; iy < this.amountY; iy++) {
        positions.push(ix * this.separation - (this.amountX * this.separation) / 2); // x
        positions.push(0); // y
        positions.push(iy * this.separation - (this.amountY * this.separation) / 2); // z

        color.setHex(0xffffff);

        colors.push(color.r, color.g, color.b);

        sizes.push(20);
      }
    }

    this.geometry.addAttribute('position', new THREE.Float32BufferAttribute(positions, 3).setDynamic(true));
    this.geometry.addAttribute('color', new THREE.Float32BufferAttribute(colors, 3));
    this.geometry.addAttribute('size', new THREE.Float32BufferAttribute(sizes, 1).setDynamic(true));

    this.particleSystem = new THREE.Points(this.geometry, this.material);
    this.scene.add(this.particleSystem);

    this.renderer = new THREE.WebGLRenderer({ alpha: true });
    this.renderer.setPixelRatio(window.devicePixelRatio);
    this.renderer.setSize(this.width, this.height);
    this.container.current.appendChild(this.renderer.domElement);

    if (this.debug) {
      this.stats = new Stats();
      // Figure out cleaner way to do this
      this.container.current.parentElement.parentElement.appendChild(this.stats.domElement);
    }

    // window.addEventListener('resize', this.onWindowResize, false);
    (function() {
      const throttle = function(type, name, obj) {
        obj = obj || window;
        let running = false;
        const func = function() {
          if (running) {
            return;
          }
          running = true;
          requestAnimationFrame(function() {
            obj.dispatchEvent(new CustomEvent(name));
            running = false;
          });
        };
        obj.addEventListener(type, func);
      };

      throttle('resize', 'optimizedResize');
    })();

    window.addEventListener('optimizedResize', this.onWindowResize);

    document.addEventListener('mouseenter', () => {
      this.hover = true;
    });
    document.addEventListener('mouseout', () => {
      this.hover = false;
    });
    document.addEventListener('mousedown', () => {
      this.jump = 1;
    });
    document.addEventListener('mousemove', this.onMouseMove);


    document.addEventListener('touchstart', () => {
      this.hover = true;
    });
    document.addEventListener('touchend', () => {
      this.hover = false;
    });

    document.addEventListener('touchmove', this.onMouseMove);
  };

  onWindowResize = () => {
    this.width = this.container.current.clientWidth;
    this.height = this.container.current.clientHeight;

    this.camera.aspect = this.width / this.height;
    let cameraPos = {
      x: this.camera.position.x,
      y: 500 + this.height * 2,
      z: 2000 + (this.height + this.width),
    };

    this.sizeMultiplier = 1 + this.height / 450;

    this.camera.position.set(cameraPos.x, cameraPos.y, cameraPos.z);
    this.camera.updateProjectionMatrix();

    this.renderer.setSize(this.width, this.height);
  };

  animate = () => {
    this.renderWaves();

    if (this.debug) this.stats.update();
    requestAnimationFrame(this.animate);
  };

  renderWaves = () => {
    if (this.hover) {
      this.camera.position.x += this.mouseX * 0.7 * 0.05 - this.camera.position.x * 0.06;
      this.camera.position.y += -(this.mouseY * 0.7) * 0.05 - this.camera.position.y * 0.07;
      this.camera.position.y += 100;
    } else {
      if (this.camera.position.x < this.originalCameraPos.x) {
        this.camera.position.x += (this.originalCameraPos.x - this.camera.position.x) * 0.02;
      } else if (this.camera.position.x > this.originalCameraPos.x) {
        this.camera.position.x -= (this.camera.position.x - this.originalCameraPos.x) * 0.02;
      }

      if (this.camera.position.y < this.originalCameraPos.y) {
        this.camera.position.y += (this.originalCameraPos.y - this.camera.position.y) * 0.02;
      } else if (this.camera.position.y > this.originalCameraPos.y) {
        this.camera.position.y -= (this.camera.position.y - this.originalCameraPos.y) * 0.02;
      }
    }

    this.camera.lookAt(this.scene.position);

    let positionPointer = 0;
    let sizePointer = 0;
    const time = this.clock.getElapsedTime() * 5;

    const sizes = this.geometry.getAttribute('size').array;
    const positions = this.geometry.getAttribute('position').array;

    for (var ix = 0; ix < this.amountX; ix++) {
      for (var iy = 0; iy < this.amountY; iy++) {
        const deltaX = ix + time;
        const deltaY = iy + time;
        positions[positionPointer + 1] = Math.sin(deltaX * 0.3) * 50 + Math.sin(deltaY * 0.5) * 50 + (this.jump ? Math.random() * this.jump * 150 : 0);
        const size = Math.cos(deltaX * 0.2) + (Math.sin(deltaY * 0.5) + 1) * (Math.cos(ix * 100) + 6) * this.sizeMultiplier;
        sizes[sizePointer] = size < 0 ? 0 : size;

        sizePointer++;
        positionPointer += 3;
      }
    }

    if (this.jump > 0) {
      this.jump -= 0.05;
    } else if (this.jump > 1) {
      this.jump = 1;
    } else if (this.jump < 0) {
      this.jump = 0;
    }

    this.geometry.getAttribute('size').needsUpdate = true;
    this.geometry.getAttribute('position').needsUpdate = true;
    this.renderer.render(this.scene, this.camera);
  };

  render() {
    return (
      <div className={styles.container}>
        <div className={styles.root} ref={this.container} />
      </div>
    );
  }
}

export default Wave;
