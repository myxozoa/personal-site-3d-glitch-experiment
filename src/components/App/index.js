import React from 'react';
import { Router } from '@reach/router';
import anime from 'animejs';

import styles from './App.module.css';

import Waves from '../Waves';
import Header from '../Header';

import Landing from '../Pages/Landing';
import Projects from '../Pages/Projects';
import Contact from '../Pages/Contact';
import NotFound from '../Pages/NotFound';

import { getMousePos } from '../../helpers/getMousePos';

class App extends React.Component {
  constructor(props) {
    super(props);

    this.header = React.createRef();
    this.content = React.createRef();
    this.bounds = null;
  }

  componentDidMount() {
    this.bounds = document.body.getBoundingClientRect();

    window.addEventListener('resize', () => (this.bounds = document.body.getBoundingClientRect()));
  }

  mouseEnter = () => {
    anime.remove(this.header.current);
    anime.remove(this.content.current);
  };

  mouseMove = event => {
    // event.preventDefault();
    const mousePos = getMousePos(event);
    const relMousePos = { x: mousePos.x - this.bounds.left, y: mousePos.y - this.bounds.top };

    const config = {
      translation: { x: [5, -5], y: [5, -5], z: [0, 0] },
      rotation: { x: [2, -2], y: [-2, 2], z: [0, 0] },
      // rotation: { x: [0, 0], y: [0, 0], z: [0, 0] }
    };

    const transforms = {
      translation: {
        x: ((config.translation.x[1] - config.translation.x[0]) / this.bounds.width) * relMousePos.x + config.translation.x[0],
        y: ((config.translation.y[1] - config.translation.y[0]) / this.bounds.height) * relMousePos.y + config.translation.y[0],
        z: 30,
      },
      rotation: {
        x: ((config.rotation.x[1] - config.rotation.x[0]) / this.bounds.height) * relMousePos.y + config.rotation.x[0],
        y: ((config.rotation.y[1] - config.rotation.y[0]) / this.bounds.width) * relMousePos.x + config.rotation.y[0],
        z: 0,
      },
    };

    for (let test in transforms.rotation) {
      if (transforms.rotation[test] > 5) {
        transforms.rotation[test] = 5;
      } else if (transforms.rotation[test] < -5) {
        transforms.rotation[test] = -5;
      }
    }


    const transform3d =
      `translate3d(${transforms.translation.x}px,${transforms.translation.y}px,${transforms.translation.z}px)` +
      ' rotateX(' +
      transforms.rotation.x +
      'deg) rotateY(' +
      transforms.rotation.y +
      'deg) rotateZ(' +
      transforms.rotation.z +
      'deg)';


    this.header.current.style.transform = transform3d;
    this.content.current.style.transform = transform3d;
  };

  mouseLeave = event => {
    anime({
      targets: this.header.current,
      duration: 1500,
      easing: 'easeOutElastic',
      elasticity: 400,
      translateX: 0,
      translateY: 0,
      translateZ: [30, 30],
      rotateX: 0,
      rotateY: 0,
      rotateZ: 0,
    });

    anime({
      targets: this.content.current,
      duration: 1500,
      easing: 'easeOutElastic',
      elasticity: 400,
      translateX: 0,
      translateY: 0,
      translateZ: [30, 30],
      rotateX: 0,
      rotateY: 0,
      rotateZ: 0,
    });
  };

  render() {
    return (
      <div
        className={styles.app}
        onMouseMove={this.mouseMove}
        onMouseLeave={this.mouseLeave}
        onMouseEnter={this.mouseEnter}
        onTouchMove={this.mouseMove}
        onTouchEnd={this.mouseLeave}
        onTouchStart={this.mouseEnter}
      >
        <header ref={this.header} className={styles.header}>
          <Header />
        </header>

        <div className={styles.content} ref={this.content}>
          <Router>
            <Landing path="/" />
            <Projects path="/projects" />
            <Contact path="/contact" />
            <NotFound default />
          </Router>
        </div>

        <Waves width="100" height="100" separation="50" amountX="120" amountY="30" />
      </div>
    );
  }
}

export default App;
