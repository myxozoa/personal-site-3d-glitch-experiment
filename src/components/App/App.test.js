import React from 'react';
import { shallow } from 'enzyme';

import Waves from '../Waves';
import App from './index';

describe('<App />', () => {
  it('renders the Waves component', () => {
    const app = shallow(<App />);

    expect(app.find(Waves)).toHaveLength(1);
  });
});