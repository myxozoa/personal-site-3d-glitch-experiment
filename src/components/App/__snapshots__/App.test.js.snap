// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<div
  className="app"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
>
  <header
    className="header"
  >
    <Header />
  </header>
  <div
    className="content"
  >
    <Router>
      <Landing
        path="/"
      />
      <Projects
        path="/projects"
      />
      <Contact
        path="/contact"
      />
      <Landing
        default={true}
      />
    </Router>
  </div>
  <Wave
    amountX="120"
    amountY="30"
    height="100"
    separation="50"
    width="100"
  />
</div>
`;
